### PR TITLE
Copter: Pre-Arm Safety Check for Auto Missions

### PR DESCRIPTION
**Pull request by rishabsingh3003**
This PR solves #1348 which prevents a **fly away** of copter. 
I have created a new param that takes in max allowed distance of the first waypoint when the user switches to mode auto. If the new param, I.e `MIS_MAX_WP_DIST` is set as '0' this feature is disabled (It is disabled by default). If the distance to first way point is greater than maximum allowed distance then copter does not switch to mode auto and raises a message alert. 

Here is what currently happens on master if a user switches to mode auto when the mission was set from some different location far away. It can easily lead to a fly away situation. 

Screenshot from 2020-01-17 18-50-28 (copy)

After these commits, and setting up the param:
'MIS_MAX_WP_DIST` to any distance in meters
and 
`ARMING_MIS_ITEMS` to value: 64 (bitmask 6), the following message is raised:

Screenshot from 2020-01-19 20-23-18

I don't know if this is the best approach to solving this problem, but I feel it'll be a good safety feature to add. 
@rmackay9 @peterbarker @magicrub , if possible, please suggest any changes that can be made to make this PR better.

### CODE REVIEW

File: ArduCopter/Parameters.cpp [hunk 1]
```diff
@@ -963,6 +963,12 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     AP_SUBGROUPINFO(arot, "AROT_", 37, ParametersG2, AC_Autorotation),
 #endif
 
+    // @Param: AUTO_MAX_WP_DIST
+    // @DisplayName: Maximum allowed distance to first waypoint in mode auto
+    // @Description: Maximum allowed distance in meters to first waypoint to allow switching to mode auto. Prevents a fly away. Set param to 0, to disable the feature. 
+    // @User: Standard
```
Comment (amilcarlucas) [inline ArduCopter/Parameters.cpp]: The units need to be specified using @Units: and a @range: must also be specified.

Comment (rishabsingh3003) [reply ArduCopter/Parameters.cpp]: @amilcarlucas I wanted to put the range there but I got a little confused on what I should write. Setting '0' disables the features and the range can be anything that `uint32_t` can store. So should I put range to soemthing like 10-10000 (10km)

**Comment by peterbarker** [comment]
You know, I could've sworn we already had this...

In any case, it would be nice to have this for Plane and Rover too!

Best way there is to add and check the parameter in AP_Mission and the main AP_Arming instead....

**Comment by peterbarker** [comment]
@rishabsingh3003 Note `AP_Arming::mission_checks` in AP_Arming/AP_Arming.cpp`

**Comment by rishabsingh3003** [comment]
@peterbarker done! Thanks for the suggestions.  The check has been moved to `AP_Arming::mission_checks` and the parameter to `AP_Mission`

File: libraries/AP_Mission/AP_Mission.h [hunk 2]
```diff
@@ -480,6 +480,11 @@ class AP_Mission {
     // returns true if the mission contains the requested items
     bool contains_item(MAV_CMD command) const;
 
+    //return maximum dist to first wp parameter
+    uint32_t get_max_arm_dist(void) {
```
Comment (peterbarker) [inline libraries/AP_Mission/AP_Mission.h]: `const`

File: libraries/AP_Mission/AP_Mission.h [hunk 3]
```diff
@@ -555,6 +560,7 @@ class AP_Mission {
     AP_Int16                _cmd_total;  // total number of commands in the mission
     AP_Int8                 _restart;   // controls mission starting point when entering Auto mode (either restart from beginning of mission or resume from last command run)
     AP_Int16                _options;    // bitmask options for missions, currently for mission clearing on reboot but can be expanded as required
+    AP_Int32                _max_nav_dist; //maximum allowed distance to first waypoint
```
Comment (peterbarker) [inline libraries/AP_Mission/AP_Mission.h]: Probably worth considering the type here.  int32 gives a rather too-large range.  Is 65km too little?

File: libraries/AP_Mission/AP_Mission.cpp [hunk 4]
```diff
@@ -31,6 +31,14 @@ const AP_Param::GroupInfo AP_Mission::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("OPTIONS",  2, AP_Mission, _options, AP_MISSION_OPTIONS_DEFAULT),
 
+    // @Param: MAX_WP_DIST
+    // @DisplayName: Maximum allowed distance to first waypoint.
+    // @Description: Maximum allowed distance in meters to first waypoint for a pre-arm check. Prevents a fly away. Set param to 0, to disable the feature. 
+    // @Units: m
+    // @Range: 10 10000
```
Comment (peterbarker) [inline libraries/AP_Mission/AP_Mission.cpp]: 10km isn't enough for a Plane.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 5]
```diff
@@ -612,6 +612,20 @@ bool AP_Arming::mission_checks(bool report)
                 return false;
             }
           }
+        AP_Mission::Mission_Command cmd;
+        //get first cmd 
+        if(mission->get_next_nav_cmd(1, cmd) && (_required_mission_items & MIS_ITEM_CHECK_NAV_DIST)) {
+            Location present_loc;
+            if (AP::ahrs().get_position(present_loc)) {
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: Should fail if we don't know where we are.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 6]
```diff
@@ -612,6 +612,20 @@ bool AP_Arming::mission_checks(bool report)
                 return false;
             }
           }
+        AP_Mission::Mission_Command cmd;
+        //get first cmd 
+        if(mission->get_next_nav_cmd(1, cmd) && (_required_mission_items & MIS_ITEM_CHECK_NAV_DIST)) {
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: Bad assumption that the second waypoint is a relevant location command.  Check `starts_with_takeoff_command` in `AP_Mission`.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 7]
```diff
@@ -104,7 +104,7 @@ const AP_Param::GroupInfo AP_Arming::var_info[] = {
     // @Param: MIS_ITEMS
     // @DisplayName: Required mission items
     // @Description: Bitmask of mission items that are required to be planned in order to arm the aircraft
-    // @Bitmask: 0:Land,1:VTOL Land,2:DO_LAND_START,3:Takeoff,4:VTOL Takeoff,5:Rallypoint
+    // @Bitmask: 0:Land,1:VTOL Land,2:DO_LAND_START,3:Takeoff,4:VTOL Takeoff,5:Rallypoint, 6:Maximum WP Distance
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: Could be "6:Maximum distance to first waypoint"

File: ArduCopter/Parameters.cpp (hunk 1 above)
Comment (amilcarlucas) [reply ArduCopter/Parameters.cpp]: something like:
`@range: 0 20000`
should do it.

**Comment by rishabsingh3003** [comment]
@peterbarker I incorporated all the changes you suggested in the updated PR. I still believe AP_INT32 is a better option for the parameter. Gives users more flexibility. But I can, of course, change it to AP_INT16 if needed. 
Please let me know if there are any other changes needed to be made.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 8]
```diff
@@ -612,6 +612,20 @@ bool AP_Arming::mission_checks(bool report)
                 return false;
             }
           }
+        AP_Mission::Mission_Command cmd;
+        //get first cmd 
+        if(mission->get_next_nav_cmd(1, cmd) && (_required_mission_items & MIS_ITEM_CHECK_NAV_DIST)) {
+            Location present_loc;
+            if (AP::ahrs().get_position(present_loc)) {
+                float dist_to_wp = present_loc.get_distance(cmd.content.location);
```
[the remark this answers is not in the document]
Comment (rishabsingh3003) [reply libraries/AP_Arming/AP_Arming.cpp]: @WickedShell I did not understand clearly what you were suggesting here. Can you please describe it a little more?

File: libraries/AP_Mission/AP_Mission.cpp [hunk 9]
```diff
@@ -31,6 +31,14 @@ const AP_Param::GroupInfo AP_Mission::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("OPTIONS",  2, AP_Mission, _options, AP_MISSION_OPTIONS_DEFAULT),
 
+    // @Param: MAX_WP_DIST
+    // @DisplayName: Maximum allowed distance to first waypoint.
+    // @Description: Maximum allowed distance in meters to first waypoint for a pre-arm check. Prevents a fly away. Set param to 0, to disable the feature. 
+    // @Units: m
+    // @Range: 10 100000
```
Comment (amilcarlucas) [inline libraries/AP_Mission/AP_Mission.cpp]: Minimum must be set to 0 because zero is a valid value

**Pull request closed without merging**